### DISCUSSION TITLE
Pass url param to react-query integration

### DISCUSSION THREAD

**Discussion by lwhiteley** [pmndrs/jotai#500] [Q&A] [answered]
Was reading some resources related to the repository around using a param to make a request but would like to know if there is any recipe for doing this with the react query plugin?

Reading references so far:
https://github.com/pmndrs/jotai/issues/68
https://codesandbox.io/s/jotai-query-hacker-news-u4sli?from-embed=&file=/src/App.tsx:1057-1064

the idea around what i want to achieve is to trigger a query doing the following

```jsx
import React, { useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { Provider, atom, useAtom } from 'jotai';
import { useAtomValue, useUpdateAtom } from 'jotai/utils';
import { atomWithQuery } from 'jotai/query';

const setting = atomWithQuery((get) => ({
// id should be an input somehow
  ...buildQueryConfig({ id }),
}));

const fetchAtom = atom(
  // unclear what should go here
);

export const Component = () => {
  const { id } = useParams();
  const [data, fetch] = useAtom(fetchAtom);

  useEffect(() => {
    fetch(id);
  }, []);


  return (
    <>
     something to display
    </>
  );
};
```

Is this the right track or is there a better way? Also advice on the unclear parts would be nice

Thanks

**Comment by dai-shi** [comment]
Hey, thanks for asking.
Assuming you want to globally define the atom with RQ, here's what I have in mind.

```js
const idAtom = atom(null)

const dataAtom = atomWithQuery((get) => {
  const id = get(idAtom)
  if (id === null) {
    return ... // OOPS: there's no way to stop initialization...
  }
  return buildQueryConfig({ id })
})

export const Component = ({ id }) => {
  const setId = useUpdateAtom(idAtom)
  useEffect(() => {
    setId(id)
  }, [setId, id])
  return <>something to display</>
}
```

So, there's a missing feature in `atomWithQuery`. Otherwise, we'd need to wrap with another atom, which is too complicated for app code.

@Aslemammad any other idea?

**Reply by lwhiteley** [reply]
ill try it.

 just a quick note regarding cannot stop initialization maybe just use the query config `enabled` prop. it could stop the request from going out. something like

```js
const dataAtom = atomWithQuery((get) => {
  const id = get(idAtom)
  return buildQueryConfig({ id, enabled: !!id })
})
```

**Reply by dai-shi** [reply]
Oh, RQ already has that capability? Nice! Please try if it works. We can still consider stopping on our end.

**Reply by lwhiteley** [reply]
Conceptually the enabled prop works i think. However, using the `<Suspense>` example, the situation is that because the request is never sent/completed then its stuck in a loading state. 

This is with your example or with the enabled prop so doesnt make a difference due to both not sending a request.

// Doesnt work
```js
const idAtom = atom(null)

const dataAtom = atomWithQuery((get) => {
  const id = get(idAtom)
  return buildQueryConfig({ id, enabled: !!id })
})

export const Component = ({ id }) => {
  const setId = useUpdateAtom(idAtom)
  useEffect(() => {
    setId(id)
  }, [setId, id])
  return <>something to display</>
}
```

However i did a second test 

// Works but causes two requests to be sent. first one is semantically bad request as its not the real item needed on first render

```js
const idAtom = atom('some-known-id') // only change made

const dataAtom = atomWithQuery((get) => {
  const id = get(idAtom)
  return buildQueryConfig({ id, enabled: !!id })
})

export const Component = ({ id }) => {
  const setId = useUpdateAtom(idAtom)
  useEffect(() => {
    setId(id)
  }, [setId, id])
  return <>something to display</>
}
```

So there needs to be some proper way to delay the. execution until the param is known

**Reply by dai-shi** [reply]
While it delays the execution, can it suspend? meaning `<Suspense>` will render the fallback.

**Reply by lwhiteley** [reply]
The fallback is being rendered indefinitely actually. due to `atomWithQuery ` not completing its job. (or at least this is my theory from my limited understanding).

The only other thing i can think of is to pass a noop `queryFn` to react-query so it renders at least once .. but this would defeat the purpose of using Suspense mode i guess  and also potentially reduce the integrity of the query

**Reply by dai-shi** [reply]
> he fallback is being rendered indefinitely actually. due to atomWithQuery not completing its job. 

Can you repro the behavior with codesandbox?

My expectation is if we `setId`, the suspense is canceled. But, as this is a new feature, there might still be a bug or a limitation.

**Reply by lwhiteley** [reply]
ok ill reproduce and send it in a few hours as i had to step out

**Reply by lwhiteley** [reply]
ok so to make things simple i used the code from your sandbox but initialized the id atom as null and tried to set it in PostTitle component.

You can see here that it gets stuck indefinitely

https://codesandbox.io/s/reverent-swirles-vgubo?file=/src/App.js

**Reply by dai-shi** [reply]
> https://codesandbox.io/s/reverent-swirles-vgubo?file=/src/App.js

For me, this works as expected. If you click the button, it will show the title later. I don't think we as a community know what to do with Suspense in general (not jotai specific).

Here's a workaround to hide the loading fallback.
https://codesandbox.io/s/wispy-bird-j9flo?file=/src/App.js

**Reply by lwhiteley** [reply]
That would still be unacceptable IMHO as it would degrade the user experience.

Reason: 
- If i were given a link for eg `myblog.com/blog/:blogid`
- I click on the link from a chat or message a friend sent me
- I arrive at the page and see a button that is saying "click me to really see the blog post"

This would be annoying and not the expectation of the consumer

> I don't think we as a community know what to do with Suspense in general (not jotai specific)

This is true, However, after reading more of the documentation and sleeping on it I came up with 1 solution/workaround and 1 suggestion.

### Solution/Workaround
https://codesandbox.io/s/condescending-gagarin-igwln?file=/src/App.js

**Description:** The solution uses `onMount` instead to initialize the postId (not ideal but works)

**Note:** 
I also tried to just move the useEffect outside the Suspense component and this did not work (Im not sure why). This may still be a bug as `setId` is actually being called and the query atom is not subscribing to the change. see here https://codesandbox.io/s/sharp-drake-b8i7u?file=/src/App.js

### Suggestion

For apps that do not use `Suspense` maybe there should be an alternative atom for this scenario. maybe `asyncAtomWithQuery` 🤷🏾‍♂️ . This would return the normal use case api for react query with `{isLoading, isIdle, data, ...}`

I tried to use the `atomWithQuery` outside Suspense and it throws an error

useful link: https://react-query.tanstack.com/guides/suspense

**Reply by dai-shi** [reply]
This is a nice discussion. Thanks for opening it up.

> That would still be unacceptable IMHO as it would degrade the user experience.
> 
> Reason:
> 
>     * If i was given a link for eg `myblog.com/blog/:blogid`
>     * I click on the link from a chat or message a friend sent me
>     * I arrive at the page and see a button that is saying "click me to really see the blog post"

Ah, I miss that part of the use case. Yeah, this makes sense. I will read about the rest later.

**Reply by dai-shi** [reply]
> ### Suggestion
> 
> For apps that do not use `Suspense` maybe there should be an alternative atom for this scenario.

This might be possible, but it doesn't solve the usage with Suspense, which is a primary usage in jotai.

We had thought about supporting non-suspense usage in `jotai/query`, but we decided to focus on Suspense.

> ### Solution/Workaround
> 
> https://codesandbox.io/s/condescending-gagarin-igwln?file=/src/App.js
> 
> **Description:** The solution uses `onMount` instead to initialize the postId (not ideal but works)

Actually, this is probably the right solution, not a workaround.
Updating id from router event makes so much sense, because doing so from effects is too late, feels wrong.
I'd like to see how you would do with the router.

> **Note:**
> I also tried to just move the useEffect outside the Suspense component and this did not work (Im not sure why). This may still be a bug as `setId` is actually being called and the query atom is not subscribing to the change. see here https://codesandbox.io/s/sharp-drake-b8i7u?file=/src/App.js

This is probably because your `useUpdateAtom` is outside the `Provider`.

**Reply by lwhiteley** [reply]
> This is probably because your useUpdateAtom is outside the Provider.

Oh wow 😄 totally missed that.

ill try it again and also a version with the router to see

**Reply by lwhiteley** [reply]
sweet now it works with the useEffect

https://codesandbox.io/s/confident-hugle-7uy9m?file=/src/App.js

ill try with the router next

**Reply by lwhiteley** [reply]
so here is my first crack at it:
https://codesandbox.io/s/interesting-swirles-r5qjl?file=/src/App.js

There is no real event to get the params as it is quite synchronous. but seems to work.

Let me know if there is a better way you have in mind

**Reply by lwhiteley** [reply]
One issue im seeing is refreshing the query..sometimes the post data is not the correct one.. looking into it

Happens after clicking "next" then i click on "home" and go back to a post. The post data doesnt update

**Reply by dai-shi** [reply]
Yes, we should listen to history, instead of useParams+useEffect, which is delayed.
Here you go:
https://codesandbox.io/s/quirky-paper-u472t?file=/src/App.js

> Happens after clicking "next" then i click on "home" and go back to a post. The post data doesnt update

It was a bug, we fixed in v0.16.8. ref: #498

**Reply by lwhiteley** [reply]
nice thanks.

in order for it to work for all routes i used a route config instead
https://codesandbox.io/s/amazing-frog-yjynf?file=/src/App.js

Update: looking at a small bug where the params are not set in similar flow
> Happens after clicking "next" then i click on "home" and go back to a post. The post data doesnt update

**Reply by lwhiteley** [reply]
issue: When setParams gets called with an empty object and then with a populated object again, it doesnt update

// Update: nope my bad..it is updating..something else is the issue

but is getting stuck again in a loading state. For some reason `atomWithQuery` is not updating even though setParams is being called
<img width="413" alt="Screenshot 2021-05-29 at 4 52 58 AM" src="https://user-images.githubusercontent.com/6108793/120055994-d7230f00-c039-11eb-9c2c-706119fb1e71.png">

**Reply by dai-shi** [reply]
can repro with v0.16.8? there's a possibility it still has a bug (like #498 wasn't enough). would be nice to have a minimal reproduction without atomWithQuery...

**Reply by lwhiteley** [reply]
would have to start with atomWithQuery to debug actually.
As you can see the`<Id />` component is updating but `atomWithquery` is not
without knowing anything deeper, these would be the facts

and yep this is with v0.16.8

**Reply by lwhiteley** [reply]
So issue: atomWithQuery is being called with an id of undefined 
Therefore the query gets disabled.
After this the atom never receives another update 

to prove this i slightly updated the code and now it works
https://codesandbox.io/s/relaxed-morse-vi4wq?file=/src/App.js

```
if (match && match.params && match.params.id) {
        console.log(match.params);
        setParams(match.params);
      }
```

However this isnt practical as there could be a param map without an id in it legitimately

**Reply by lwhiteley** [reply]
The other issue here is that i could have multiple queries listening to the global setParams change. So i think the better approach is to find a way to know when to accept changes as well. 

meaning only listen on the routes that need the query

Update: would only listen when mounted. however the undefined happens when already mounted + changing a route so would need to filter this action somehow

**Reply by lwhiteley** [reply]
unfortunately ill have to go back to the solution in the below sandbox (slightly updated) for now as having one listener has too many edge case right now.
Also it feels like im rewriting the routing logic

https://codesandbox.io/s/dark-pond-q9qyo?file=/src/App.js

ill maybe give it another go when i have more time.

**Reply by dai-shi** [reply]
I'm not 100% following, but if the issue is about depending on `enabled`, I have an idea to fix.

**Reply by lwhiteley** [reply]
As usual slept on it.

To make this more bearable, I gave the responsibility to the route to say if the atom should be updated

https://codesandbox.io/s/relaxed-morse-vi4wq?file=/src/App.js:4602-4622 

snippet:
```js
routes: [
      {
        path: "/posts/:id",
        component: Post,
        shouldUpdateAtom: ({ match }) =>
          match && match.params && match.params.id
      }
    ]
```

**Reply by dai-shi** [reply]
So, do we still have any issue that should be fixed on the library end, or not?

**Reply by lwhiteley** [reply]
But note: 
still doesnt solve the unmount issue. (Navigating to a route that doesnt depend on the query and the param is undefined )

There are still issues with this:
- if i navigate to a route that also has shouldUpdateAtom and it doesnt have an id param then it will end up in the same situation
- For a production solution we would have to have highly specific param names like `:postId` as a generic thing like `:id` can cause conflicts for other queries when the unmount behaviour happens (or scope params per route)

**Reply by lwhiteley** [reply]
> So, do we still have any issue that should be fixed on the library end, or not?

- I think it would still be handy to have a way to not get stuck when a query is disabled. that would need to be fixed anyway. This is the main blocker really

**Reply by dai-shi** [reply]
I see this seems like a bug in the library. I'm still not sure it's atomWithQuery only or even without it.

**Reply by dai-shi** [reply]
@lwhiteley Can you help me by creating a minimal reproduction of the bug in codesandbox with atomWithQuery without react-router?

**Reply by lwhiteley** [reply]
Yep I'll send it to you later when I get some time

**Reply by lwhiteley** [reply]
Hi @dai-shi 

Here is a simple reproduction to showcase the issue.

https://codesandbox.io/s/jotai-query-hacker-news-forked-xmf0z?file=/src/App.tsx

First click hide and then show, here you will see the atom gets stuck

**Reply by dai-shi** [reply]
https://codesandbox.io/s/jotai-query-hacker-news-forked-3h5vg
Thanks. I slimmed it a little more. I'm investigating it now.

(update: i'm struggling with it. there must be something fundamental. creating various tests.)

**Reply by lwhiteley** [reply]
hmm interesting. Let me know if you make any strides

**Comment by lwhiteley** [comment, marked as answer]
As a side note, I started looking into jotai as a way to manage a client heavy state component for an app that already uses react-query.

What interested me was the react-query integration and also the recipe for managing large objects. it is pretty neat

With that i had to ensure that only one query client was being used so this is what i did

```js
import React, { useEffect } from 'react';
import { QueryClientProvider, QueryClient } from 'react-query';
import { queryClientAtom } from 'jotai/query';
import { useUpdateAtom } from 'jotai/utils';

const queryClient = new QueryClient();

export const JotaiQueryProvider = ({ children }) => {
  const setQueryClient = useUpdateAtom(queryClientAtom);

  useEffect(() => {
    setQueryClient(queryClient);
  }, [setQueryClient]);

  return (
    <>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </>
  );
};
```

May be helpful for someone else i guess

**Reply by dai-shi** [reply]
You can use Jotai's `Provider` to initialize queryClient.
```jsx
<Provider initialValues={[[queryClientAtom, queryClient]]}>
```
This should be better than `useEffect`.

**Reply by lwhiteley** [reply]
ah sweet much better

**Reply by dai-shi** [reply]
I've made a fix. #505 
Would you try the codesandbox build in the PR?
https://github.com/pmndrs/jotai/pull/505#issuecomment-854537646 -> Click `here` -> See "Local install instructions"

**Reply by lwhiteley** [reply]
Awesome I'll check it today

**Reply by lwhiteley** [reply]
Sorry for late follow up but yes!! it works as expected 🚀 .

sanbox: https://codesandbox.io/s/jotai-query-hacker-news-forked-s79go?file=/src/App.tsx

Thanks for getting this resolved so quickly 👍🏾 

is there more cleanup to do and when will this be released?

**Reply by dai-shi** [reply]
It's already released, but I'll cut another release soon and that includes some refactoring to this.

**Reply by dai-shi** [reply]
Published: https://www.npmjs.com/package/jotai/v/0.16.10

**Reply by dai-shi** [reply]
> It's already released

it was wrong. v0.16.10 is the first version that has the fix.

**Reply by lwhiteley** [reply]
Cool thanks

**Comment by aulneau** [comment]
This might be helpful! https://github.com/pmndrs/jotai/discussions/535

**Comment by Mayandev** [comment]
How about multiple params
